crates/hyle-model/src/contract.rs

```diff
@@ -315,3 +316,41 @@ impl Add<BlockHeight> for BlockHeight {
         BlockHeight(self.0 + other.0)
     }
 }
+
+#[derive(Debug, Serialize, Deserialize, Default, Clone, PartialEq, Eq, Encode, Decode)]
+pub struct RegisterContractAction {
+    pub verifier: Verifier,
+    pub program_id: ProgramId,
+    pub state_digest: StateDigest,
+    pub contract_name: ContractName,
+}
+
+impl Hashable<TxHash> for RegisterContractAction {
+    fn hash(&self) -> TxHash {
+        let mut hasher = Sha3_256::new();
+        hasher.update(self.verifier.0.clone());
+        hasher.update(self.program_id.0.clone());
+        hasher.update(self.state_digest.0.clone());
+        hasher.update(self.contract_name.0.clone());
+        let hash_bytes = hasher.finalize();
+        TxHash(hex::encode(hash_bytes))
+    }
+}
+
+impl ContractAction for RegisterContractAction {
+    fn as_blob(
+        &self,
+        contract_name: ContractName,
+        caller: Option<BlobIndex>,
+        callees: Option<Vec<BlobIndex>>,
+    ) -> Blob {
+        Blob {
+            contract_name,
+            data: BlobData::from(StructuredBlobData {
```

Comment:
On veut vraiment utiliser un StructuredBlobData ici ? Je vois pas l'usage

Reply:
IMHO on devrait avoir 2 traits: 

```rust
trait StructuredBlobContractAction {
    fn as_blob(
        &self,
        contract_name: ContractName,
        caller: Option<BlobIndex>,
        callees: Option<Vec<BlobIndex>>,
    ) -> Blob;
}

trait SimpleBlobContractAction {
    fn as_blob(&self, contract_name: ContractName) -> Blob;
}
```

mais je sais pas où ce trait "ContractAction" est requis, on peut peut-être pas faire ça en fait

Reply:
J'ai utilisé structuredBlob parce que comme ça si quelqu'un veut faire une fonction qui appelle "register_contract" dans un autre contrat c'est compatible et ça me semble worth it (et staking fait ça je crois).

Idéalement on pourrait être compatible avec ou sans caller, mais la façon dont on a fait les structuredblob le permet pas en l'état.

Reply:
> une fonction qui appelle "register_contract" dans un autre contrat c'est compatible

je comprends pas :thinking: 

dans staking j'ai utilisé un Structuredblob en prévision des opérations de claim rewards si je me souviens bien

Reply:
Je veux dire tu peux envisager un contrat qui appelle un contrat pour register un nouveau domain-name, et l'identité c'est le caller contract, pas `identity`. Avoir les caller/callees ça me semble worth it